### ISSUE TITLE
[Cache]: Get invalid cache items

### ISSUE THREAD

**Issue by BlackbitNeueMedien** [symfony/symfony#27294]
**Description**  
It would be nice of one could get invalid cache items of a tag-based cache adapter to be able to warmup a cache after some items got invalid.

**Example**  
Database query and its result get saved as serialized object together with tag 'test'. Due to a change all cache items with tag 'test' get invalid (via TagAwareAdapter::invalidateTags(['test'])). Later on it would be nice to fetch all invalid entries to re-warmup the cache by executing the queries and fill the cache with the current values.

**Comment by nicolas-grekas** [comment]
I've been wondering about something like that also. Right now, I feel like doing it with "graced" items might create a high amount of complexity, requiring new methods to fetch items from graced pools, being able to deal with a mix of hard and soft deleted items (should we create a set of interfaces/methods to allow controlling hard/soft delete, etc.)?

I've recently proposed two PRs that are related:
- #27009 to add probabilistic early expiration. This works for expired items, but not for invalidated tags.
- #27028 to add per fpm pool locking, so that synchronous recomputation happens only once per front server.

I'm wondering if these proposed approches couldn't be enough in practice to make it unecessary to implement anything like graced items. That'd be great because that'd mean less complexity.
WDYT?

**Comment by BlackbitNeueMedien** [comment]
I think I found a way: 

1. Initially set cache expiration to 0 (as I do not know when a write operation will occur which could alter the result)
1. When a write operation occurs, fetch all cache items by tag which could have been altered. Set their expiration date to past.
1. In a separate process fetch all cache items and check whether they have been expired or not and re-execute database query to re-warm the cache.

Would this be possible? Especially I wonder if 2. is possible (get cache items by tag).

**Comment by nicolas-grekas** [comment]
2. is not possible, and I think it goes off topic actually. Being able to do that requires a relational system, which does not exist and poses scalability issues. Solvable for sure, but not in a generic way IMHO.
Btw, *if* you happen to be able to know which items relate to tag "foo", another way to do it is by making tag invalidation do *nothing* BUT send a message to another process that could then recompute items in the background. This way, no need for graced items at all, and things might be optimal as everything happens in the background.

**Comment by BlackbitNeueMedien** [comment]
> another way to do it is by making tag invalidation do nothing BUT send a message to another process
You mean by implementing an own class which implements TagAwareAdapterInterface?

**Comment by nicolas-grekas** [comment]
> implementing an own class which implements TagAwareAdapterInterface

correct!

**Comment by BlackbitNeueMedien** [comment]
I underestimated your "if you happen to be able to know which items relate to a tag"

Is it possible to let AdapterInterface::getItems() return all items, regardless whether they are valid or not? Invalidated items could get isHit = false. We could introduce a second parameter $invalid = false to AdapterInterface::getItems() so there would not be a backwards incomptible change.

**Comment by nicolas-grekas** [comment]
I think it's not: design wise having a well though but restricted domain is desired. E.g. fetching all items from the backend doesn't scale with millions of items...

**Comment by BlackbitNeueMedien** [comment]
I meant not all items in general but all items which have the given tag, e.g. AdapterInterface::getItems(['foo']) could return CacheItems with isHit=true (valid items) and with isHit=false (invalid items).

**Comment by nicolas-grekas** [comment]
> items which have the given key or tags

that still exposes to the scalability issue and requires storing the relation in the backend.
Varnish (which is a great example of a tagged backend) doesn't have such feature.
That's why I don't this belongs to the caching layer.

**Comment by nicolas-grekas** [comment]
Closing as "won't fix", as explained. Thanks for proposing.
